File: academic-system-python/academic_system/model/academic_system.py
from typing import List, Optional

from .academic_class import AcademicClass


class AcademicSystem:
    _instance: Optional["AcademicSystem"] = None

    def __init__(self):
        self._classes: List[AcademicClass] = []

    # Retorna a única instância do sistema, criando-a na primeira chamada (Singleton)
    @classmethod
    def get_instance(cls) -> "AcademicSystem":
        if cls._instance is None:
            cls._instance = AcademicSystem()
        return cls._instance

    # Reseta o Singleton — usado principalmente nos testes para isolar o estado
    @classmethod
    def reset(cls) -> None:
        cls._instance = None

    # Adiciona uma nova turma à lista interna
    def add_class(self, academic_class: AcademicClass) -> None:
        self._classes.append(academic_class)

    # Busca e retorna uma turma pelo código, ou None se não encontrada
    def find_by_code(self, code: str) -> Optional[AcademicClass]:
        for c in self._classes:
            if c.code == code:
                return c
        return None

    # Retorna uma cópia da lista de turmas para evitar modificações externas
    def get_classes(self) -> List[AcademicClass]:
        return list(self._classes)

    # Substitui todas as turmas pelas fornecidas (usada ao trocar tipo de persistência)
    def replace_classes(self, classes: List[AcademicClass]) -> None:
        self._classes = list(classes)

File: academic-system-java/src/main/java/org/example/academic/system/model/AcademicSystem.java
package org.example.academic.system.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

public class AcademicSystem {
    private final List<AcademicClass> classes;

    // Inicializa o sistema com uma lista vazia de turmas
    public AcademicSystem() {
        this.classes = new ArrayList<>();
    }

    // Adiciona uma nova turma à lista do sistema
    public void addClass(AcademicClass academicClass) {
        classes.add(academicClass);
    }

    // Substitui todas as turmas existentes pelas turmas carregadas (usada ao trocar tipo de persistência)
    public void replaceClasses(List<AcademicClass> loadedClasses) {
        classes.clear();

        if (loadedClasses != null) {
            classes.addAll(loadedClasses);
        }
    }

    // Busca uma turma pelo código, ignorando maiúsculas/minúsculas e espaços
    public Optional<AcademicClass> findClassByCode(String code) {
        if (code == null) {
            return Optional.empty();
        }

        return classes.stream()
                .filter(academicClass -> academicClass.getCode().equalsIgnoreCase(code.trim()))
                .findFirst();
    }

    // Retorna a lista de turmas como visão imutável para evitar modificações externas
    public List<AcademicClass> getClasses() {
        return Collections.unmodifiableList(classes);
    }
}
